File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])
        directions = [[-1, 0], [0, -1], [1, 0], [0, 1]]

        def dfs(i: int, j: int, word_index: int) -> bool:
            if word_index == len(word) - 1:
                return True
            
            tmp = board[i][j]
            board[i][j] = '.'

            for di, dj in directions:
                ni, nj = i + di, j + dj
                if 0 <= ni < m and 0 <= nj < n and board[ni][nj] == word[word_index + 1]:
                    if dfs(ni, nj, word_index + 1):
                        return True
            
            board[i][j] = tmp
            return False

        for i in range(m):
            for j in range(n):
                if board[i][j] == word[0]:
                    if dfs(i, j, 0):
                        return True
        
        return False

File: 0079-word-search/0079-word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if(backtracking(board, i, j, word, 0)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean backtracking(char[][] board, int i, int j, String word, int index) {
        if (index == word.length()) {
            return true;
        }

        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || board[i][j] != word.charAt(index)) {
            return false;
        }

        char tmp = board[i][j];
        board[i][j] = '#'; // 为了不让接下来的查找走回头路
        boolean ret = backtracking(board, i + 1, j, word, index + 1)
                    || backtracking(board, i - 1, j, word, index + 1)
                    || backtracking(board, i, j + 1, word, index + 1)
                    || backtracking(board, i, j - 1, word, index + 1);
        board[i][j] = tmp;
        return ret;
    }
}